dss/stepfunctions/visitation/__init__.py

```diff
@@ -0,0 +1,295 @@
+
+        setattr(cls, 'sentinel_state_spec', {
+            ** mcs.base_state_spec,
+            ** mcs.sentinel_state_spec,
+            ** additional_sentinel_state_spec
+        })
+
+        return cls
+
+
+class Visitation(metaclass=VisitationStateMeta):
+    """
+    Base class vor AWS Step Function sentinel-workers datastore batch processing. This is meant to serve as a highly
+    parallelized, high throughput architecture to visit blobs in the datastore and perform generic processing.
+
+    Although Visitation is somewhat specialized for datastore processing, subclasses may largely override the propagated
+    state and behaviour of the sentinel and walker step functions, hijacking the paralell architecture for
+    other purposes.
+
+    subclasses should be registered in registered_visitations to make them available to the sentinel and walker step
+    functions.
+    """
+
+    def __init__(self, state_spec: dict, state: dict, logger) -> None:
+        """
+        Pull in fields defined in state specifications and set as instance properties
+        """
+        self.state_spec = state_spec
+
+        for k, default in state_spec.items():
+            v = state.get(k, None)
+
+            if v is None:
+                if callable(default):
+                    v = default()
+                else:
+                    v = default
+
+            setattr(self, k, v)
+
+        self.logger = logger
+
+    def get_state(self) -> dict:
+        """
+        Return step function state at the end of each lambda defined in the sentinel and walker step functions.
+        """
+        return {
+            k: getattr(self, k)
+            for k in self.state_spec
+        }
+
+    @classmethod
+    def with_sentinel_state(cls, state: dict, logger) -> 'Visitation':
+        """
+        Pull in state specific to the sentinel.
+        """
+        state_spec = cls.sentinel_state_spec
+        return cls(state_spec, state, logger)
+
+    @classmethod
+    def with_walker_state(cls, state: dict, logger) -> 'Visitation':
+        """
+        Pull in state specific to the walkers.
+        """
+        state_spec = cls.walker_state_spec
+        return cls(state_spec, state, logger)
+
+    @classmethod
+    def start(cls, replica: str, bucket: str, k_workers: int) -> str:
+        name = '{}--{}'.format(cls.__name__, str(uuid4()))
+
+        inp = {
+            'visitation_class_name': cls.__name__,
+            'name': name,
+            'replica': replica,
+            'bucket': bucket,
+            'k_workers': k_workers,
+        }
+
+        step_functions_invoke('dss-visitation-sentinel-{stage}', name, inp)
+
+        return name
+
+    def sentinel_muster(self) -> typing.List[str]:
+        """
+        Start walkers as needed based on pool of prefixes that are waiting to be executed. When all walkers are
+        finished, sets state function status code to SUCCEEDED
+        """
+
+        if self.waiting is None:  # type: ignore
```

Comment:
What are these type annotation for?

Reply:
I don't understand how to keep mypy happy without them.

Reply:
Interesting. Let me try.